# Machine Learning - Hands on Lab - Session #1

* **Lecturer:** Jonathan DEKHTIAR
* **Date:** 2017-03-13
<br/><br/>
* **Contact:** [contact@jonathandekhtiar.eu](mailto:contact@jonathandekhtiar.eu)
* **Twitter:** [@born2data](https://twitter.com/born2data)
* **LinkedIn:** [JonathanDEKHTIAR](https://fr.linkedin.com/in/jonathandekhtiar)
* **Personal Website:** [JonathanDEKHTIAR](http://www.jonathandekhtiar.eu)
* **RSS Feed:** [FeedCrunch.io](https://www.feedcrunch.io/@dataradar/)
* **Tech. Blog:** [born2data.com](http://www.born2data.com/)
* **Github:** [DEKHTIARJonathan](https://github.com/DEKHTIARJonathan)
<br/><br/>

```
*************************************************************************
**
** 2017 March 13
**
** In place of a legal notice, here is a blessing:
**
**    May you do good and not evil.
**    May you find forgiveness for yourself and forgive others.
**    May you share freely, never taking more than you give.
**
*************************************************************************
```

## 1. Loading the Python libraries

In [1]:
import os
from datetime import datetime

import numpy as np
import pandas as pd

import sklearn as sk

## 2. Re-Import the Check NaN Function from Part 2

In [2]:
def check_NaN_Values_in_df(df):
    # searching for NaN values is all the columns
    for col in df:
        nan_count = df[col].isnull().sum()

        if nan_count != 0:
            print (col + " => "+  str(nan_count) + " NaN Values")

## 3. Loading in the Data from Session #2

In [3]:
df_all = pd.read_csv(
    "output/cleaned.csv", 
    dtype={
        'country_destination': str
    }
)

# We transform again the date column into datetime
df_all['date_account_created'] = pd.to_datetime(df_all['date_account_created'], format='%Y-%m-%d %H:%M:%S')
df_all['timestamp_first_active'] = pd.to_datetime(df_all['timestamp_first_active'], format='%Y-%m-%d %H:%M:%S')

# Check for NaN Values => We must find: country_destination => 62096 NaN Values
check_NaN_Values_in_df(df_all) 

df_all.sample(n=5) # Only display a few lines and not the whole dataframe

country_destination => 62096 NaN Values


affiliate_channel affiliate_provider  age country_destination  \
189335            direct             direct   45                 NaN   
14098             direct             direct   -1                 NDF   
59195      sem-non-brand             google   21                  US   
174921            direct             direct   22                 NaN   
171378         sem-brand             google   20                 NaN   

       date_account_created first_affiliate_tracked first_browser  \
189335           2014-08-19                  linked            IE   
14098            2013-04-23               untracked        Chrome   
59195            2013-10-18                     omg        Chrome   
174921           2014-07-29                     omg        Safari   
171378           2014-07-25                     omg        Safari   

       first_device_type     gender          id language signup_app  \
189335    Android Tablet     FEMALE  wqk4w0iu6q       en        Web   
14098        Mac Desktop  -unknown-  pxbt7xbyeo       en        Web   
59195        Mac Desktop     FEMALE  4q6pcdllzf       en        Web   
174921       Mac Desktop     FEMALE  z1gdq6qft3       en        Web   
171378       Mac Desktop       MALE  3nl1gjku5u       en        Web   

        signup_flow signup_method timestamp_first_active  
189335            0         basic    2014-08-19 04:11:28  
14098             0         basic    2013-04-23 21:37:17  
59195             0         basic    2013-10-18 07:19:27  
174921            0         basic    2014-07-29 21:36:38  
171378            0         basic    2014-07-25 01:19:48

## 3. Transforming Categorical Data

Let's go for some **One Hot Encoding** - replacing the categorical fields in the dataset with multiple columns representing one value from each column.

In [4]:
# Home made One Hot Encoding function
def convert_to_binary(df, column_to_convert):
    categories = list(df[column_to_convert].drop_duplicates())

    for category in categories:
        cat_name = str(category).replace(" ", "_").replace("(", "").replace(")", "").replace("/", "_").replace("-", "").lower()
        col_name = column_to_convert[:5] + '_' + cat_name[:10]
        df[col_name] = 0
        df.loc[(df[column_to_convert] == category), col_name] = 1

    return df

In [5]:
columns_to_convert = [
    'gender', 
    'signup_method', 
    'signup_flow', 
    'language', 
    'affiliate_channel', 
    'affiliate_provider', 
    'first_affiliate_tracked', 
    'signup_app', 
    'first_device_type', 
    'first_browser'
]

# One Hot Encoding
for column in columns_to_convert:
    df_all = convert_to_binary(df=df_all, column_to_convert=column)
    df_all.drop(column, axis=1, inplace=True)
    
df_all.sample(n=5)

age country_destination date_account_created          id  \
15451    45                 NDF           2013-04-30  rxg661asyo   
183421   48                 NaN           2014-08-10  d691g89xdu   
59604    27                  US           2013-10-20  lbrqsxv55b   
185505   63                 NaN           2014-08-13  s0uxw0vm4d   
132328   -1                 NDF           2014-05-18  kneo0ttza4   

       timestamp_first_active  gende_other  gende_female  gende_male  \
15451     2013-04-30 16:16:56            0             1           0   
183421    2014-08-10 21:22:54            0             1           0   
59604     2013-10-20 06:33:53            0             0           1   
185505    2014-08-13 16:55:21            0             0           0   
132328    2014-05-18 00:42:19            0             0           0   

        gende_unknown  signu_basic        ...         first_crazy_brow  \
15451               0            1        ...                        0   
183421              0            0        ...                        0   
59604               0            0        ...                        0   
185505              1            0        ...                        0   
132328              1            1        ...                        0   

        first_stainless  first_coolnovo  first_opera_mini  first_googlebot  \
15451                 0               0                 0                0   
183421                0               0                 0                0   
59604                 0               0                 0                0   
185505                0               0                 0                0   
132328                0               0                 0                0   

        first_outlook_20  first_icedragon  first_ibrowse  first_nintendo_b  \
15451                  0                0              0                 0   
183421                 0                0              0                 0   
59604                  0                0              0                 0   
185505                 0                0              0                 0   
132328                 0                0              0                 0   

        first_uc_browser  
15451                  0  
183421                 0  
59604                  0  
185505                 0  
132328                 0  

[5 rows x 140 columns]

In [6]:
# Add new date related fields
df_all['day_account_created'] = df_all['date_account_created'].dt.weekday
df_all['month_account_created'] = df_all['date_account_created'].dt.month
df_all['quarter_account_created'] = df_all['date_account_created'].dt.quarter
df_all['year_account_created'] = df_all['date_account_created'].dt.year
df_all['hour_first_active'] = df_all['timestamp_first_active'].dt.hour
df_all['day_first_active'] = df_all['timestamp_first_active'].dt.weekday
df_all['month_first_active'] = df_all['timestamp_first_active'].dt.month
df_all['quarter_first_active'] = df_all['timestamp_first_active'].dt.quarter
df_all['year_first_active'] = df_all['timestamp_first_active'].dt.year
df_all['created_less_active'] = (df_all['date_account_created'] - df_all['timestamp_first_active']).dt.days

# Drop unnecessary columns
columns_to_drop = ['date_account_created', 'timestamp_first_active', 'date_first_booking', 'country_destination']
for column in columns_to_drop:
    if column in df_all.columns:
        df_all.drop(column, axis=1, inplace=True)

print ("Dataframe Shape:", df_all.shape)
df_all.sample(n=5)

Dataframe Shape: (216973, 147)


age          id  gende_other  gende_female  gende_male  gende_unknown  \
111004   26  056qnsw0nz            0             1           0              0   
51235    -1  wx274oia10            0             0           0              1   
157908   -1  3owd2169i2            0             0           0              1   
147463   -1  eqb26ysarh            0             0           1              0   
41371    46  1lz0uzwk2t            0             1           0              0   

        signu_basic  signu_facebook  signu_google  signu_weibo  \
111004            1               0             0            0   
51235             1               0             0            0   
157908            1               0             0            0   
147463            1               0             0            0   
41371             0               1             0            0   

               ...           day_account_created  month_account_created  \
111004         ...                             6                      3   
51235          ...                             5                      9   
157908         ...                             6                      7   
147463         ...                             1                      6   
41371          ...                             4                      8   

        quarter_account_created  year_account_created  hour_first_active  \
111004                        1                  2014                 19   
51235                         3                  2013                 20   
157908                        3                  2014                 18   
147463                        2                  2014                 17   
41371                         3                  2013                 21   

        day_first_active  month_first_active  quarter_first_active  \
111004                 6                   3                     1   
51235                  5                   9                     3   
157908                 6                   7                     3   
147463                 1                   6                     2   
41371                  4                   8                     3   

        year_first_active  created_less_active  
111004               2014                   -1  
51235                2013                   -1  
157908               2014                   -1  
147463               2014                   -1  
41371                2013                   -1  

[5 rows x 147 columns]

## 5. Mid - Conclusion

We had **14 Columns**, we now have **147 columns**. 

It seems like a lot, however we mostly just have **restructured** the information and only created **10 columns**


## 6. Adding new data

### 6.1. Understanding session.csv
Exactly like we did with the training / testing data. We now investigate session data.

In [7]:
df_sessions = pd.read_csv("../input/sessions.csv")
print ("DF Session Shape:", df_sessions.shape)
df_sessions.head(n=5) # Only display a few lines and not the whole dataframe

DF Session Shape: (10567737, 6)


user_id          action action_type        action_detail  \
0  d1mm9tcy42          lookup         NaN                  NaN   
1  d1mm9tcy42  search_results       click  view_search_results   
2  d1mm9tcy42          lookup         NaN                  NaN   
3  d1mm9tcy42  search_results       click  view_search_results   
4  d1mm9tcy42          lookup         NaN                  NaN   

       device_type  secs_elapsed  
0  Windows Desktop         319.0  
1  Windows Desktop       67753.0  
2  Windows Desktop         301.0  
3  Windows Desktop       22141.0  
4  Windows Desktop         435.0

### 6.2. Cleaning and Transforming the Data

#### 6.2.1. Extract the primary and secondary devices for each user

The first piece of information we are going to extract is the primary and secondary device for each user. 

How do we determine what is the user's primary and secondary devices are? We look at how much time they spent on each device.

In [8]:
# Determine primary device
sessions_device = df_sessions.loc[:, ['user_id', 'device_type', 'secs_elapsed']]
aggregated_lvl1 = sessions_device.groupby(['user_id', 'device_type'], as_index=False, sort=False).aggregate(np.sum)
idx = aggregated_lvl1.groupby(['user_id'], sort=False)['secs_elapsed'].transform(max) == aggregated_lvl1['secs_elapsed']
df_primary = pd.DataFrame(aggregated_lvl1.loc[idx , ['user_id', 'device_type', 'secs_elapsed']])
df_primary.rename(columns = {'device_type':'primary_device', 'secs_elapsed':'primary_secs'}, inplace=True)
df_primary = convert_to_binary(df=df_primary, column_to_convert='primary_device')
df_primary.drop('primary_device', axis=1, inplace=True)

df_primary.sample(n=5)

user_id  primary_secs  prima_windows_de  prima_mac_deskto  \
152649  dyke4kcixz     1718805.0                 0                 0   
69911   xnmisg1wss     2502621.0                 1                 0   
94067   gjleobm010     1102337.0                 0                 0   
27591   ff73gwanb3      777888.0                 1                 0   
67149   51qqdt2tno      421168.0                 0                 0   

        prima_iphone  prima_ipad_table  prima_unknown  prima_android_ap  \
152649             1                 0              0                 0   
69911              0                 0              0                 0   
94067              1                 0              0                 0   
27591              0                 0              0                 0   
67149              0                 0              0                 0   

        prima_linux_desk  prima_tablet  prima_chromebook  prima_android_ph  \
152649                 0             0                 0                 0   
69911                  0             0                 0                 0   
94067                  0             0                 0                 0   
27591                  0             0                 0                 0   
67149                  0             0                 0                 1   

        prima_ipodtouch  prima_blackberry  prima_windows_ph  prima_opera_phon  
152649                0                 0                 0                 0  
69911                 0                 0                 0                 0  
94067                 0                 0                 0                 0  
27591                 0                 0                 0                 0  
67149                 0                 0                 0                 0

In [9]:
# Determine Secondary device
remaining = aggregated_lvl1.drop(aggregated_lvl1.index[idx])
idx = remaining.groupby(['user_id'], sort=False)['secs_elapsed'].transform(max) == remaining['secs_elapsed']
df_secondary = pd.DataFrame(remaining.loc[idx , ['user_id', 'device_type', 'secs_elapsed']])
df_secondary.rename(columns = {'device_type':'secondary_device', 'secs_elapsed':'secondary_secs'}, inplace=True)
df_secondary = convert_to_binary(df=df_secondary, column_to_convert='secondary_device')
df_secondary.drop('secondary_device', axis=1, inplace=True)

df_secondary.sample(n=5)

user_id  secondary_secs  secon_unknown  secon_android_ph  \
69009   on397vk5l2       1532331.0              0                 0   
87534   t1t9wc9c6o        925598.0              1                 0   
169143  2mya54ytjs         79337.0              0                 0   
40241   3zwrrfpkik        229510.0              0                 0   
3628    gkuj0jzjf0        143963.0              0                 0   

        secon_ipad_table  secon_android_ap  secon_mac_deskto  secon_iphone  \
69009                  0                 0                 1             0   
87534                  0                 0                 0             0   
169143                 0                 0                 1             0   
40241                  0                 0                 1             0   
3628                   0                 0                 0             1   

        secon_windows_de  secon_linux_desk  secon_tablet  secon_blackberry  \
69009                  0                 0             0                 0   
87534                  0                 0             0                 0   
169143                 0                 0             0                 0   
40241                  0                 0             0                 0   
3628                   0                 0             0                 0   

        secon_windows_ph  secon_chromebook  secon_opera_phon  secon_ipodtouch  
69009                  0                 0                 0                0  
87534                  0                 0                 0                0  
169143                 0                 0                 0                0  
40241                  0                 0                 0                0  
3628                   0                 0                 0                0

#### 6.2.2. Determine Counts of Actions

The next thing we are going to do is take counts of how many times each action was taken by each user. This is a two-step process. 

To handle the multiple action columns, we repeat these steps for each column individually, effectively creating three separate tables. 

Because we have now created tables where each row represents one user, we can now **join these three tables together** on the basis of the user id.

In [10]:
# Count occurrences of value in a column
def convert_to_counts(df, id_col, column_to_convert):
    id_list = df[id_col].drop_duplicates()
    
    df_counts = df.loc[:,[id_col, column_to_convert]]
    df_counts['count'] = 1
    df_counts = df_counts.groupby(by=[id_col, column_to_convert], as_index=False, sort=False).sum()
    
    new_df = df_counts.pivot(index=id_col, columns=column_to_convert, values='count')
    new_df = new_df.fillna(0)
    
    # Rename Columns
    categories = list(df[column_to_convert].drop_duplicates())
    for category in categories:
        cat_name = str(category).replace(" ", "_").replace("(", "").replace(")", "").replace("/", "_").replace("-", "").lower()
        col_name = column_to_convert + '_' + cat_name
        new_df.rename(columns = {category:col_name}, inplace=True)
        
    return new_df

In [11]:
# Aggregate and combine actions taken columns

session_actions = df_sessions.loc[:,['user_id', 'action', 'action_type', 'action_detail']]
columns_to_convert = ['action', 'action_type', 'action_detail']

session_actions = session_actions.fillna('not provided')
first = True

for column in columns_to_convert:
    print("Converting " + column + " column...")
    current_data = convert_to_counts(df=session_actions, id_col='user_id', column_to_convert=column)

    # If first loop, current data becomes existing data, otherwise merge existing and current
    if first:
        first = False
        actions_data = current_data
    else:
        actions_data = pd.concat([actions_data, current_data], axis=1, join='inner')
        
actions_data.sample(n=5)

Converting action column...
Converting action_type column...
Converting action_detail column...


action_10  action_11  action_12  action_15  action_about_us  \
user_id                                                                   
bxddkg1jtk        0.0        0.0        0.0        0.0              0.0   
x4k7l6k06i        0.0        0.0        0.0        0.0              0.0   
603ny3606m        0.0        0.0        0.0        0.0              0.0   
80pr8b4ifj        0.0        0.0        0.0        0.0              0.0   
7denusbceb        0.0        0.0        0.0        0.0              0.0   

            action_accept_decline  action_account  \
user_id                                             
bxddkg1jtk                    0.0             0.0   
x4k7l6k06i                    0.0             0.0   
603ny3606m                    0.0             0.0   
80pr8b4ifj                    0.0             0.0   
7denusbceb                    0.0             0.0   

            action_acculynk_bin_check_failed  \
user_id                                        
bxddkg1jtk                               0.0   
x4k7l6k06i                               0.0   
603ny3606m                               0.0   
80pr8b4ifj                               0.0   
7denusbceb                               0.0   

            action_acculynk_bin_check_success  action_acculynk_load_pin_pad  \
user_id                                                                       
bxddkg1jtk                                0.0                           0.0   
x4k7l6k06i                                0.0                           0.0   
603ny3606m                                0.0                           0.0   
80pr8b4ifj                                0.0                           0.0   
7denusbceb                                0.0                           0.0   

                      ...             action_detail_view_resolutions  \
user_id               ...                                              
bxddkg1jtk            ...                                        0.0   
x4k7l6k06i            ...                                        0.0   
603ny3606m            ...                                        0.0   
80pr8b4ifj            ...                                        0.0   
7denusbceb            ...                                        0.0   

            action_detail_view_search_results  \
user_id                                         
bxddkg1jtk                              126.0   
x4k7l6k06i                                0.0   
603ny3606m                               26.0   
80pr8b4ifj                                9.0   
7denusbceb                                0.0   

            action_detail_view_security_checks  \
user_id                                          
bxddkg1jtk                                 0.0   
x4k7l6k06i                                 0.0   
603ny3606m                                 0.0   
80pr8b4ifj                                 0.0   
7denusbceb                                 0.0   

            action_detail_view_user_real_names  action_detail_wishlist  \
user_id                                                                  
bxddkg1jtk                                 0.0                     0.0   
x4k7l6k06i                                 0.0                     0.0   
603ny3606m                                 0.0                     0.0   
80pr8b4ifj                                 0.0                     3.0   
7denusbceb                                 0.0                     0.0   

            action_detail_wishlist_content_update  \
user_id                                             
bxddkg1jtk                                    0.0   
x4k7l6k06i                                    1.0   
603ny3606m                                   17.0   
80pr8b4ifj                                    1.0   
7denusbceb                                    0.0   

            action_detail_wishlist_note  action_detail_your_listings  \
user_id                                                     

#### 6.2.3. Combine Data Sets

The final steps are to combine the various datasets we have created into one large dataset. 

1. we combine the two device dataframes (df_primary and df_secondary) to create a device dataframe
2. we combine the device dataframe with the actions dataframe to create a sessions dataframe with all the features we extracted from sessions.csv
3. Finally, we combine the sessions dataframe with the training and testing data dataframe

In [12]:
# Merge device datasets
df_primary.set_index('user_id', inplace=True)
df_secondary.set_index('user_id', inplace=True)
device_data = pd.concat([df_primary, df_secondary], axis=1, join="outer")

# Merge device and actions datasets
combined_results = pd.concat([device_data, actions_data], axis=1, join='outer')
df_sessions = combined_results.fillna(0)

# Merge user and session datasets
df_all.set_index('id', inplace=True)
df_all = pd.concat([df_all, df_sessions], axis=1, join='inner')

df_all.sample(n=5)

age  gende_other  gende_female  gende_male  gende_unknown  \
9ck8a5k8os   -1            0             0           0              1   
vqc4gn02p1   34            0             0           0              1   
0uwkt2xutm   -1            0             0           0              1   
j9jh1z5rl5   -1            0             0           0              1   
xdqhbtya7z   -1            0             0           0              1   

            signu_basic  signu_facebook  signu_google  signu_weibo  signu_24  \
9ck8a5k8os            1               0             0            0         0   
vqc4gn02p1            1               0             0            0         0   
0uwkt2xutm            1               0             0            0         0   
j9jh1z5rl5            1               0             0            0         0   
xdqhbtya7z            1               0             0            0         0   

                      ...             action_detail_view_resolutions  \
9ck8a5k8os            ...                                        0.0   
vqc4gn02p1            ...                                        0.0   
0uwkt2xutm            ...                                        0.0   
j9jh1z5rl5            ...                                        0.0   
xdqhbtya7z            ...                                        0.0   

            action_detail_view_search_results  \
9ck8a5k8os                                0.0   
vqc4gn02p1                                0.0   
0uwkt2xutm                                2.0   
j9jh1z5rl5                               16.0   
xdqhbtya7z                                7.0   

            action_detail_view_security_checks  \
9ck8a5k8os                                 0.0   
vqc4gn02p1                                 0.0   
0uwkt2xutm                                 0.0   
j9jh1z5rl5                                 0.0   
xdqhbtya7z                                 0.0   

            action_detail_view_user_real_names  action_detail_wishlist  \
9ck8a5k8os                                 0.0                     0.0   
vqc4gn02p1                                 0.0                     0.0   
0uwkt2xutm                                 0.0                     0.0   
j9jh1z5rl5                                 0.0                     0.0   
xdqhbtya7z                                 0.0                     0.0   

            action_detail_wishlist_content_update  \
9ck8a5k8os                                    0.0   
vqc4gn02p1                                    0.0   
0uwkt2xutm                                    0.0   
j9jh1z5rl5                                    6.0   
xdqhbtya7z                                    0.0   

            action_detail_wishlist_note  action_detail_your_listings  \
9ck8a5k8os                          0.0                          0.0   
vqc4gn02p1                          0.0                          0.0   
0uwkt2xutm                          0.0                          0.0   
j9jh1z5rl5                          0.0                          0.0   
xdqhbtya7z                          0.0                          0.0   

            action_detail_your_reservations  action_detail_your_trips  
9ck8a5k8os                              0.0                       0.0  
vqc4gn02p1                              0.0                       0.0  
0uwkt2xutm                              0.0                       0.0  
j9jh1z5rl5                              0.0                       0.0  
xdqhbtya7z                              0.0                       0.0  

[5 rows x 703 columns]

In [13]:
df_sessions = df_sessions.astype(int)
df_sessions.sample(n=5)

primary_secs  prima_windows_de  prima_mac_deskto  prima_iphone  \
jqrbmpact6        338454                 1                 0             0   
kjx1gpmx4j        366466                 0                 0             0   
3uz5di2cy2       2088724                 0                 0             1   
gry1r8n6ej       1086544                 0                 0             0   
ksxgqoa4ne       1783745                 0                 1             0   

            prima_ipad_table  prima_unknown  prima_android_ap  \
jqrbmpact6                 0              0                 0   
kjx1gpmx4j                 0              0                 0   
3uz5di2cy2                 0              0                 0   
gry1r8n6ej                 1              0                 0   
ksxgqoa4ne                 0              0                 0   

            prima_linux_desk  prima_tablet  prima_chromebook  \
jqrbmpact6                 0             0                 0   
kjx1gpmx4j                 0             0                 0   
3uz5di2cy2                 0             0                 0   
gry1r8n6ej                 0             0                 0   
ksxgqoa4ne                 0             0                 0   

                      ...             action_detail_view_resolutions  \
jqrbmpact6            ...                                          0   
kjx1gpmx4j            ...                                          0   
3uz5di2cy2            ...                                          0   
gry1r8n6ej            ...                                          0   
ksxgqoa4ne            ...                                          0   

            action_detail_view_search_results  \
jqrbmpact6                                  0   
kjx1gpmx4j                                  0   
3uz5di2cy2                                  0   
gry1r8n6ej                                  4   
ksxgqoa4ne                                  0   

            action_detail_view_security_checks  \
jqrbmpact6                                   0   
kjx1gpmx4j                                   0   
3uz5di2cy2                                   0   
gry1r8n6ej                                   0   
ksxgqoa4ne                                   0   

            action_detail_view_user_real_names  action_detail_wishlist  \
jqrbmpact6                                   0                       0   
kjx1gpmx4j                                   0                       0   
3uz5di2cy2                                   0                       0   
gry1r8n6ej                                   0                       0   
ksxgqoa4ne                                   0                       0   

            action_detail_wishlist_content_update  \
jqrbmpact6                                      1   
kjx1gpmx4j                                      0   
3uz5di2cy2                                      0   
gry1r8n6ej                                      1   
ksxgqoa4ne                                      0   

            action_detail_wishlist_note  action_detail_your_listings  \
jqrbmpact6                            0                            0   
kjx1gpmx4j                            0                            0   
3uz5di2cy2                            0                            0   
gry1r8n6ej                            0                            0   
ksxgqoa4ne                            0                            0   

            action_detail_your_reservations  action_detail_your_trips  
jqrbmpact6                                0                         0  
kjx1gpmx4j                                0                         0  
3uz5di2cy2                                0                         0  
gry1r8n6ej                                0                         2  
ksxgqoa4ne                                0                         1  

[5 rows x 557 columns]

In [14]:
#Just recheck we don't have new NaN Values before saving our data
# We must find no NaN, the column "country_destination" has been deleted

check_NaN_Values_in_df(df_all) 

## 7. Saving the DataFrame to csv

In [15]:
# We create the output directory if necessary
if not os.path.exists("output"):
    os.makedirs("output")
    
# We export to csv
df_all.to_csv("output/enriched.csv", sep=',')